Starting an awesome-trio list

jay-tuckey opened the pull request:
Making a page right after the tutorial with some links to libraries that have trio support.

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,57 @@
+However, to get much useful work done you will want to use some of
+the great libraries that support trio-flavoured concurrency. This list
+is not complete, but gives a starting point. Another great way to find
+trio-compatible libraries is to search on PyPI for the `Framework :: trio`
```
The single-backticks here are causing the CI build to fail: https://travis-ci.org/python-trio/trio/jobs/600224047

I think you want double-backticks, because ReST is annoying like that.

Also, let's capitalize Trio.

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,57 @@
+* `trio-click <https://github.com/python-trio/trio-click>`__ - Python composable command line utility, trio-compatible version
+* `triopg <https://github.com/python-trio/triopg>`__ - PostgreSQL client for Trio based on asyncpg
+* `trio-mysql <https://github.com/python-trio/trio-mysql>`__ - Pure Python MySQL Client
+* `hypothesis-trio <https://github.com/python-trio/hypothesis-trio>`__ - Hypothesis plugin for trio
```
I'm actually not sure what the status is of anything below `trio-click` here... maybe just file them under categories, instead of "core" libraries?

`trio-click` → "Building command line apps"
`triopg`, `trio-mysql` → already listed under databases
`hypothesis-trio` → "Testing"

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,57 @@
+----------------------
+
+* `asks <https://github.com/theelous3/asks>`__ - asks is an async requests-like http library
+* `trustme <https://github.com/python-trio/trustme>`__ - #1 quality TLS certs while you wait, for the discerning tester
```
`trustme` should also go under "Testing"

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,57 @@
+---------------------------
+* `anyio <https://github.com/agronholm/anyio>`__ - AnyIO is a asynchronous compatibility API that allows applications and libraries written against it to run unmodified on asyncio, curio and trio.
+* `tractor <https://github.com/goodboy/tractor>`__ - tractor is an attempt to bring trionic structured concurrency to distributed multi-core Python.
+* `sniffio <https://github.com/python-trio/sniffio>`__ - This is a tiny package whose only purpose is to let you detect which async library your code is running under.
```
Maybe we should have a category for "Trio/Asyncio interoperability", listing `trio-asyncio` ("use asyncio libraries from a trio app"), `anyio`, `sniffio`?

njsmith replied:
And `tractor` could go under like... "Multi-core/multiprocessing"?

I feel like "Utility libraries" would be more like `async-generator` (Makes async generators work (better) on Python 3.5 and Python 3.6), `tricycle`, `trio-util`

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,57 @@
+* `asks <https://github.com/theelous3/asks>`__ - asks is an async requests-like http library
+* `trustme <https://github.com/python-trio/trustme>`__ - #1 quality TLS certs while you wait, for the discerning tester
+* `trio-websocket <https://github.com/HyperionGray/trio-websocket>`__ - This library implements the WebSocket protocol, striving for safety, correctness, and ergonomics.
+* `quart-trio <https://gitlab.com/pgjones/quart-trio/>`__ - `Quart <https://gitlab.com/pgjones/quart>`__ is a Python ASGI web microframework with the same API as Flask and Quart-Trio is an extension for Quart to support the Trio event loop.
```
We should list `hypercorn` too (ASGI server with first-class Trio support)

njsmith replied:
In fact how about:

* `quart-trio` – Like Flask, but for Trio. A simple and powerful framework for building async web applications and REST APIs. Tip: this is an ASGI-based framework, so you'll also need an HTTP server with ASGI support.
* `hypercorn` – An HTTP server for hosting your ASGI apps. Supports HTTP/1.1, HTTP/2, HTTP/3, and Websockets. Can be run as a standalone server, or embedded in a larger Trio app. Use it with `quart-trio`, or any other Trio-compatible ASGI framework.

jay-tuckey commented:
Thanks @njsmith , I'll update this afternoon when I'm done work.

njsmith commented:
Test failure is #1231. I'll close/re-open to retry.

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,64 @@
+
+.. currentmodule:: trio
+
+You have completed the turorial, and are enthusiastic about building
```
```suggestion
You have completed the tutorial, and are enthusiastic about building
```

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,64 @@
+However, to get much useful work done you will want to use some of
+the great libraries that support Trio-flavoured concurrency. This list
+is not complete, but gives a starting point. Another great way to find
+Trio-compatible libraries is to search on PyPI for the ``Framework :: trio``
```
```suggestion
Trio-compatible libraries is to search on PyPI for the ``Framework :: Trio``
```

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,64 @@
+* `triopg <https://github.com/python-trio/triopg>`__ - PostgreSQL client for Trio based on asyncpg
+* `trio-mysql <https://github.com/python-trio/trio-mysql>`__ - Pure Python MySQL Client
+* `sqlalchemy_aio <https://github.com/RazerM/sqlalchemy_aio>`__ - Add asyncio and Trio support to SQLAlchemy core, derived from alchimia.
+* `sphinxcontrib-trio <https://github.com/python-trio/sphinxcontrib-trio>`__ - Make Sphinx better at documenting Python functions and methods
```
`sphinxcontrib-trio` doesn't have anything to do with databases :-). 

Also probably worth mentioning that it makes it easy to document async functions.

njsmith commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,64 @@
+-----------------------------
+* `anyio <https://github.com/agronholm/anyio>`__ - AnyIO is a asynchronous compatibility API that allows applications and libraries written against it to run unmodified on asyncio, curio and trio
+* `sniffio <https://github.com/python-trio/sniffio>`__ - This is a tiny package whose only purpose is to let you detect which async library your code is running under
+* `trio-asyncio <https://github.com/python-trio/trio-asyncio>`__ - Trio-Asyncio is a re-implementation of the asyncio mainloop on top of Trio
```
While that's true, I think people will get a better idea of why they should care about trio-asyncio if we describe it as "Trio-Asyncio lets you use asyncio libraries from your Trio app."

smurfix replied:
**many** asyncio libraries. Not all of them, aiohttp being the most glaring counterexample.

njsmith commented:
@jay-tuckey Hey, just wanted to check in and see if you're still interested in working on this – I think it's pretty close, but also I know sometimes stuff comes up :-)

jay-tuckey commented:
Hi @njsmith , sorry about the delay. I got caught up in work and a busy life, but I've gone through and updated the PR now with the changes as discussed. I also fixed a few small typos and some consistency. 

Your suggestions and @smurfix are great, keep them coming. :slightly_smiling_face:

smurfix commented:
Addition: distmqtt https://github.com/smurfix/distmqtt.git which is a mostly-reasonable MQTT client and broker (an anyio-ified fork of hbmqtt).

energizah commented:
The Urwid TUI library has Trio support now https://github.com/urwid/urwid/pull/372/files

goodboy commented on docs/source/awesome-trio-libraries.rst:
```diff
@@ -0,0 +1,73 @@
+* `trio-click <https://github.com/python-trio/trio-click>`__ - Python composable command line utility, trio-compatible version.
+
+
+Multi-Core/Multiprocessing
```
Might be worth tossing [`trio_run_in_process`](https://github.com/ethereum/trio-run-in-process) here as well.

parity3 commented:
This is a constantly moving target so we should just merge-then-iterate right? A few opinions:

- would like to see triopg/trio-mysql only under Database Libraries and removed from Core Trio Libraries. If opinions differ, then where is sqlalchemy_aio in core?
- there should probably be an awesome-anyio as well?
- trio_asyncio should probably be on the list?

dhirschfeld commented:
Such a list has been requested a number of times:
* https://gitter.im/python-trio/general?at=5e18e017c395034902789127
* https://gitter.im/python-quart/lobby?at=5e12013f4e9ea12051bc4b19
* https://gitter.im/python-trio/general?at=5a615d06ae53c159030726d7

As @parity3 says, this list will never be complete - there will (hopefully) always be new, awesome trio projects being developed so I think it would be great to get an initial version in and then new libraries can be added as new PRs on an incremental, ongoing basis.

dhirschfeld commented:
@jay-tuckey, thanks for kicking this off! I think what we have here is a great start so if there are no more projects you'd like to add now it would be good to remove the `[WIP]` so we can get this merged and then continually iterate/update.

jay-tuckey commented:
Ok, I have removed the [WIP] as suggested, and also added a few more of the libraries people have mentioned.

pquentin commented:
Thanks! Merged. The PyPI search does not appears to include too many results but we can fix that in another pull request.
